examples/arty/xdc-plugin/xdc.cc

```diff
@@ -0,0 +1,212 @@
+			case INTERNAL_VREF:
+				process_vref(std::vector<std::string>(args.begin() + 2, args.end()), design);
+				break;
+			default:
+				assert(false);
+		}
+	}
+	void process_vref(std::vector<std::string> args, RTLIL::Design* design)
+       	{
+		if (args.size() != 2) {
+			log("Incorrect number of arguments: %zu\n", args.size());
+			return;
+		}
+		if (current_iobank == 0) {
+			log("set_property: No valid bank set. Use get_iobanks.\n");
+			return;
+		}
+		RTLIL::Module* top_module = design->top_module();
+		if (!design->has(ID(BANK))) {
+			RTLIL::Module* bank_module = design->addModule(ID(BANK));
+			bank_module->makeblackbox();
+			bank_module->avail_parameters.insert(ID(NUMBER));
+			bank_module->avail_parameters.insert(ID(INTERNAL_VREF));
+		}
+		char bank_cell_name[16];
+		snprintf(bank_cell_name, 16, "\\bank_cell_%d", current_iobank);
+		int internal_vref = 1000 * std::atof(args[0].c_str());
+		if (internal_vref != 600 &&
+				internal_vref != 675 &&
+				internal_vref != 750 &&
+				internal_vref != 900) {
+			log("Incorrect INTERNAL_VREF value\n");
+			return;
+		}
+		RTLIL::Cell* bank_cell = top_module->cell(RTLIL::IdString(bank_cell_name));
+		if (!bank_cell) {
+			bank_cell = top_module->addCell(RTLIL::IdString(bank_cell_name), ID(BANK));
+		}
+		bank_cell->setParam(ID(NUMBER), RTLIL::Const(current_iobank));
+		bank_cell->setParam(ID(INTERNAL_VREF), RTLIL::Const(internal_vref));
+	}
+
+} SetProperty;
+
+
+struct WriteFasm : public Backend {
+	//static std::unordered_map<int, std::string> bank_tiles;
+	WriteFasm() : Backend("fasm", "Write out FASM features") { }
+	void help() YS_OVERRIDE
+	{
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("    write_fasm filename\n");
+		log("\n");
+		log("Write out a file with vref FASM features\n");
+		log("\n");
+	}
+	void execute(std::ostream *&f, std::string filename,  std::vector<std::string> args, RTLIL::Design *design) YS_OVERRIDE
+	{
+		size_t argidx = 1;
+		extra_args(f, filename, args, argidx);
+		process_vref(f, design);
+	}
+
+	void process_vref(std::ostream *&f, RTLIL::Design* design) {
+		RTLIL::Module* top_module(design->top_module());
+		if (top_module == nullptr) {
+			log("No top module detected\n");
+			return;
+		}
+		// Return if no BANK module exists as this means there are no cells
+		if (!design->has(ID(BANK))) {
+			return;
+		}
+		// Generate a fasm feature associated with the INTERNAL_VREF value per bank
+		// e.g. VREF value of 0.675 for bank 34 is associated with tile HCLK_IOI3_X113Y26
+		// hence we need to emit the following fasm feature: HCLK_IOI3_X113Y26.VREF.V_675_MV
+		for (auto cell : top_module->cells()) {
+                        if (cell->type != ID(BANK)) continue;
+			int bank_number(cell->getParam(ID(NUMBER)).as_int());
+			int bank_vref(cell->getParam(ID(INTERNAL_VREF)).as_int());
+			*f << "HCLK_IOI3_" << bank_tiles[bank_number] <<".VREF.V_" << bank_vref << "_MV\n";
```

Comment:
Need error checking on the bank number.  E.g. if bank number not found in bank_tiles, generate error message.